pkg/crc/machine/generate_bundle.go

```diff
@@ -0,0 +1,132 @@
+package machine
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/code-ready/crc/pkg/crc/cluster"
+	"github.com/code-ready/crc/pkg/crc/constants"
+	"github.com/code-ready/crc/pkg/crc/logging"
+	"github.com/code-ready/crc/pkg/crc/machine/bundle"
+	"github.com/code-ready/crc/pkg/crc/oc"
+	crcssh "github.com/code-ready/crc/pkg/crc/ssh"
+	"github.com/code-ready/machine/libmachine/state"
+	"github.com/pkg/errors"
+)
+
+func (client *client) GenerateBundle() error {
+	bundleMetadata, sshRunner, err := loadVM(client)
+	if err != nil {
+		return err
+	}
+	defer sshRunner.Close()
+
+	ocConfig := oc.UseOCWithSSH(sshRunner)
+	if err := cluster.RemovePullSecretFromCluster(ocConfig, sshRunner); err != nil {
+		return errors.Wrap(err, "Error loading bundle metadata")
+	}
+
+	// Stop the cluster
+	currentState, err := client.Stop()
+	if err != nil {
+		return err
+	}
+	if currentState != state.Stopped {
+		return fmt.Errorf("VM is not stopped, current state is %s", currentState.String())
+	}
+
+	tmpBaseDir, err := ioutil.TempDir("", "crc_custom_bundle")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpBaseDir)
+
+	// Create the custom bundle directory which is used as top level directory for tarball during compression
+	customBundleName := bundle.GetCustomBundleName(bundleMetadata.GetBundleName())
+
+	copier, err := bundle.NewCopier(bundleMetadata, tmpBaseDir, customBundleName)
+	if err != nil {
+		return err
+	}
+	defer copier.Cleanup() //nolint
+	customBundleDir := copier.CachedPath()
+
+	if err := copier.CopyKubeConfig(); err != nil {
+		return err
+	}
+
+	kubePassword, err := cluster.GetKubeadminPassword(bundleMetadata)
+	if err != nil {
+		return err
+	}
+	if err := copier.SetKubeAdminPassword(kubePassword); err != nil {
+		return err
+	}
+
+	if err := copier.CopyPrivateSSHKey(constants.GetPrivateKeyPath()); err != nil {
+		return err
+	}
+
+	if err := copier.CopyOcExecutable(); err != nil {
+		return err
+	}
+
+	// Copy disk image
+	logging.Infof("Copy the disk image to %s", customBundleName)
+	logging.Debugf("Absolute path of custom bundle directory: %s", customBundleDir)
+	diskPath, diskFormat, err := copyDiskImage(customBundleDir, bundleMetadata.GetDiskImagePath())
+	if err != nil {
+		return err
+	}
+
+	if err := copier.SetDiskImage(diskPath, diskFormat); err != nil {
+		return err
+	}
+
+	if err := copier.GenerateBundle(customBundleName); err != nil {
+		return err
+	}
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	logging.Infof("Bundle is generated in %s", filepath.Join(cwd, customBundleName))
+	logging.Infof("You need to perform 'crc delete' and 'crc start -b %s' to use this bundle", filepath.Join(cwd, customBundleName))
```

Comment:
Regarding https://github.com/code-ready/crc/pull/2147#issuecomment-824973933
Maybe `crc delete` can remove `~/.crc/cache/crc_libvirt_custom_4.7.5/` when a custom bundle is being used? This means no need to unpack the bundle with every start.

Reply:
>Maybe crc delete can remove ~/.crc/cache/crc_libvirt_custom_4.7.5/ when a custom bundle is being used? This means no need to unpack the bundle with every start.

@cfergeau Even then you need to unpack it every start if delete remove the custom bundle right?

Reply:
If delete removes the custom bundle directory, then `crc start -b custom_bundle; crc stop; crc start; crc stop` will only unpack once. Next time you run `crc delete` then yes the bundle will need to be unpacked again.